File: C++Countdown/Round6/scorer.cpp
#include <algorithm>
#include <cctype>
#include <fstream>
#include <iostream>
#include <iomanip>
#include <vector>

using namespace std;

#include "tokens.cpp"

static void print_compiler_version()
{
    cout << ">>> Compiler is G++ " << __VERSION__ << '\n';
    cout << ">>> Standard C++ " << __cplusplus << '\n';
    cout << ">>>\n";
}

static int line_size(const string & line)
{
    int size = 0;
    for (auto & ch : line)
        if (!isspace(ch))
            size++;
    return size;
}

static int lines_size(const vector<string> & lines)
{
    int size = 0;
    for (auto & line : lines)
        size += line_size(line);
    return size;
}

static void print_program_size(const vector<string> & lines)
{
    const int width = 60;
    cout << endl;
    cout << "-----|" << string(width,'-') << endl;
    int total_size = 0;
    for (auto & line : lines)
    {
        const int size = line_size(line);
        cout << setw(3) << setfill(' ') << size << "  |" << line << endl;
        total_size += size;
    }
    cout << "-----|" << string(width, '-') << endl;
    cout << setw(3) << setfill(' ') << total_size
         << " == countdown.cpp.size" << endl;
    cout << endl;
}

// - - - - - - - - - - - - - - - - - - - -

static vector<string> read_lines(const char * filename)
{
    vector<string> lines;
    ifstream is(filename);
    string line;
    while (getline(is, line))
        lines.push_back(line);
    return lines;
}

// - - - - - - - - - - - - - - - - - - - -

static bool uses(const vector<string> & lines, const string & token)
{
    // ... also matches .
    // double also matches do
    // etc etc

    for (auto & line : lines)
        if (line.find(token) != string::npos)
            return true;
    return false;
}

static int tokens_size(const vector<string> & lines)
{
    int size = 0;
    for (auto & token : tokens)
        if (uses(lines, token))
            size += token.size();

    return size;
}

static bool missing_tokens(const vector<string> & lines)
{
    vector<string> unused;
    for (auto & token : tokens)
        if (!uses(lines, token))
            return true;

    return false;
}

// - - - - - - - - - - - - - - - - - - - -

static void print_token_bonuses(const vector<string> & lines)
{
    const int width = 20;
    int tokens_size = 0;
    cout << "-----|" << string(width,'-') << endl;
    for (auto & token : tokens)
        if (uses(lines, token))
        {
            cout << setw(3) << setfill(' ')  << token.size() << "  |" << token << endl;
            tokens_size += token.size();
        }

    for (auto & token : tokens)
        if (!uses(lines, token))
            cout << setw(3) << setfill(' ') << 0 << "  |" << token << endl;

    cout << "-----|" << string(width,'-') << endl;
    cout << setw(3) << setfill(' ') << tokens_size << " == used_tokens.size" << endl;
    int completion_bonus = missing_tokens(lines) ? 0 : 50;
    cout << setw(3) << setfill(' ') << completion_bonus << " == completion.bonus" << endl;
}

// - - - - - - - - - - - - - - - - - - - -

int main(int, const char * argv[])
{
    print_compiler_version();
    vector<string> lines = read_lines(argv[1]);
    int program_size = lines_size(lines);
    int used_token_bonus = tokens_size(lines);
    int completion_bonus = missing_tokens(lines) ? 0 : 50;

    cout << ">>> Score = -countdown.cpp.size + 3*used_tokens.size + completion.bonus" << endl
         << ">>>       = " << setw(3) << setfill(' ') << -program_size << " + "
                              << "3*" << used_token_bonus << " + "
                              << completion_bonus << endl
         << ">>>       = " << (-program_size + (3*used_token_bonus) + completion_bonus) << endl;

    cout << endl;
    print_token_bonuses(lines);
    cout << endl;
    print_program_size(lines);

    // green-traffic light pattern...put it out of sight
    for(int i = 0; i < 100; i++)
        cout << endl;
    cout << "All tests passed\n";
}

File: JavaCountdown/PracticeRound/Scorer.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;

public class Scorer {

    int lineSize(String line) {
        char[] chars = line.toCharArray();
        int size = 0;
        for (int i = 0; i < chars.length; i++) {
            if (!isSpace(chars[i]))
                size++;
        }
        return size;
    }

    boolean isSpace(char aChar) {
        return String.valueOf(aChar).trim().isEmpty();
    }

    int linesSize(List<String> lines) {
        int size = 0;
        for (String line : lines) {
            size += lineSize(line);
        }
        return size;
    }


    String hr(int width) {
        StringBuffer buf = new StringBuffer();

        buf.append("------|");
        for (int i = 0; i < width; i++) {
            buf.append("-");
        }

        return buf.toString();
    }

    void printProgramSize(List<String> lines, PrintStream out) {

        int totalSize = 0;

        out.println(hr(60));

        for (String line : lines) {
            int size = lineSize(line);

            out.printf(" %3d  |%s\n", size, line);
            totalSize += size;
        }

        out.println(hr(60));

        out.println("   " + totalSize + " == CountDown.java.size");
    }

// - - - - - - - - - - - - - - - - - - - -

    List<String> readLines(String filename) {
        List<String> lines = new ArrayList<>();

        try {
            BufferedReader reader = new BufferedReader(new InputStreamReader(getClass().getResourceAsStream(filename)));
            String sCurrentLine;
            while ((sCurrentLine = reader.readLine()) != null) {
                lines.add(sCurrentLine);
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        return lines;
    }

// - - - - - - - - - - - - - - - - - - - -

    boolean uses(List<String> lines, String token) {
        // ... also matches .
        // double also matches do
        // etc etc

        for (String line : lines) {
            if (line.contains(token)) {
                return true;
            }
        }

        return false;
    }

    int tokensSize(List<String> lines) {
        int size = 0;
        for (String token : Tokens.tokens) {
            if (uses(lines, token)) {
                size += token.length();
            }
        }

        return size;
    }

    boolean missingTokens(List<String> lines) {
        for (String token : Tokens.tokens) {
            if (!uses(lines, token)) {
                return true;
            }
        }

        return false;
    }

// - - - - - - - - - - - - - - - - - - - -

    void printTokenBonuses(List<String> lines, PrintStream out) {
        int tokensSize = 0;
        out.println(hr(20));

        for (String token : Tokens.tokens) {
            if (uses(lines, token)) {
                out.printf(" %3d  |%s\n", token.length(), token);
                tokensSize += token.length();
            }
        }

        for (String token : Tokens.tokens) {
            if (!uses(lines, token)) {
                out.printf(" %3d  |%s\n", 0, token);
            }
        }

        out.println(hr(20));

        out.printf(" %3d == used_tokens.size\n", tokensSize);

        int completion_bonus = missingTokens(lines) ? 0 : 100;

        out.printf(" %3d == completion.bonus\n", completion_bonus);
    }

// - - - - - - - - - - - - - - - - - - - -

    public void score(String filename, PrintStream out) {
        score(readLines(filename), out);

        // green-traffic light pattern...put it out of sight
        for (int i = 0; i < 100; i++)
            out.println();
        out.print("OK (1 test)");
    }

    public void score(List<String> lines, PrintStream out) {
        int program_size = linesSize(lines);
        int used_token_bonus = tokensSize(lines);
        int completion_bonus = missingTokens(lines) ? 0 : 100;

        out.println(">>> Score = -CountDown.java.size + 5*usedTokens.size + completion.bonus");
        out.println(">>>       = " + (-program_size) + " + 5*" + used_token_bonus + " + " + completion_bonus);
        out.println(">>>       = " + (-program_size + (5 * used_token_bonus) + completion_bonus));

        out.println();
        printTokenBonuses(lines, out);
        out.println();
        printProgramSize(lines, out);
    }

    public static void main(String[] args) {
        new Scorer().score(args[0], System.out);
    }

}